### PR TITLE
[dsd] Support container ID field from dogstatsd 1.2

### PR DESCRIPTION
**Pull request by ahmed-mez**
### What does this PR do?

Extend origin detection on by supporting a new field in the dsd protocol, the new field `c:<id>` holds the container ID discovered by the client.

### Motivation

- More granular tags (container tags instead of pod tags)
- Not limited to Kubernetes, and can be a good alternative in environments where origin detection on UDS is not an option (e.g Fargate)

### Describe how to test/QA your changes

Deploy the agent with `DD_DOGSTATSD_ORIGIN_DETECTION_CLIENT=true` 

```
echo -n "testing.dsd.container:1|g|#foo:bar,dd.internal.card:high|c:<container id>" | nc -u -w1 <Agent IP> 8125
```

You should be able to see container tags attached to the `testing.dsd.container` metric on the app.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -497,6 +497,7 @@ func InitConfig(config Config) {
 	// contexts will be deleted (see 'dogstatsd_expiry_seconds').
 	config.BindEnvAndSetDefault("dogstatsd_context_expiry_seconds", 300)
 	config.BindEnvAndSetDefault("dogstatsd_origin_detection", false) // Only supported for socket traffic
+	config.BindEnvAndSetDefault("dogstatsd_origin_detection_udp", false)
```
Comment (ahmed-mez) [inline pkg/config/config.go:500]: not totally convinced by this config name, suggestions are welcome!

Comment (vickenty) [reply pkg/config/config.go:500]: The setting would also enable the new field on UDS traffic. I think we want that, but the name should probably reflect this. `dogstatsd_origin_detection_full`, or perhaps `dogstatsd_origin_detection: client`.

File: pkg/dogstatsd/enrich.go [hunk 2]
```diff
@@ -42,19 +43,23 @@ func extractTagsMetadata(tags []string, defaultHostname string, originTags strin
...
 		// Check if the value is not "none" in order to avoid calling
 		// the tagger for entity that doesn't exist.
 
 		// currently only supported for pods
-		k8sOrigin = kubelet.KubePodTaggerEntityPrefix + entityIDValue
+		k8sOrigin = kubelet.KubePodTaggerEntityPrefix + originFromTag
+	} else if originFromTag == "" && originFromMsg != "" {
+		// originFromMsg is the container id sent by the newer clients.
+		// Opt-in is handled when parsing.
+		k8sOrigin = containers.BuildTaggerEntityName(originFromMsg)
```
Comment (vickenty) [inline pkg/dogstatsd/enrich.go]: Perhaps this should be re-named `originFromClient` too, to match how we call this value elsewhere.

File: pkg/dogstatsd/parse.go [hunk 3]
```diff
@@ -193,6 +207,11 @@ func (p *parser) parseFloat64List(rawFloats []byte) ([]float64, error) {
 	return values, nil
 }
 
+// extractContainerID parses the value of the container ID field.
+func (p *parser) extractContainerID(rawContainerIDField []byte) string {
+	return string(rawContainerIDField[len(containerIDFieldPrefix):])
```
Comment (vickenty) [inline pkg/dogstatsd/parse.go:212]: Not a big problem, but perhaps we could keep containerID as a `[]byte` here, until we concatenate it with the prefix. This will avoid an intermediate copy of container id.

File: pkg/dogstatsd/enrich.go (hunk 2 above)
Comment (ahmed-mez) [reply pkg/dogstatsd/enrich.go]: yes, it was done in the second commit 👍

File: pkg/config/config.go (hunk 1 above)
Comment (ahmed-mez) [reply pkg/config/config.go:500]: I'd say `dogstatsd_origin_detection_type: client`

File: releasenotes/notes/dsd-container-b2039d8e0e510fbf.yaml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+---
+enhancements:
+  - |
+    Dogstatsd origin detection on UDP supports container entities in addition to Kubernetes Pods.
```
Comment (maycmlee) [inline releasenotes/notes/dsd-container-b2039d8e0e510fbf.yaml:4]:
```suggestion
    DogstatsD origin detection on UDP supports container entities in addition to Kubernetes Pods.
```

**Review by maycmlee — APPROVED** [review]
Small edit.

File: releasenotes/notes/dsd-container-b2039d8e0e510fbf.yaml [hunk 5]
```diff
@@ -0,0 +1,4 @@
+---
+enhancements:
+  - |
+    DogstatsD origin detection on UDP supports container entities in addition to Kubernetes Pods.
```
Comment (hush-hush) [inline releasenotes/notes/dsd-container-b2039d8e0e510fbf.yaml:4]: Could we rephrase this to not talk about UDP but about supporting a new field in dogstatsd (the new field is unrelated to UDP) ?

File: pkg/config/config_template.yaml [hunk 6]
```diff
@@ -1526,6 +1526,14 @@ api_key:
 #
 # dogstatsd_origin_detection: false
 
+## @param dogstatsd_origin_detection_type - string - optional - default: ""
+## @env DD_DOGSTATSD_ORIGIN_DETECTION_TYPE - string - optional - default: ""
+## A DogStatsD client can send its container ID.
+## Set this parameter to "client" to instruct the Agent to use the container ID to enrich the metrics with container tags.
+## Note: This requires using a compatible DogStatsD client version.
```
Comment (hush-hush) [inline pkg/config/config_template.yaml:1533]:
```suggestion
## Note: This requires using a client compatible with DogStatsD protocol version 1.2.
```

File: pkg/config/config_template.yaml [hunk 7]
```diff
@@ -1526,6 +1526,14 @@ api_key:
 #
 # dogstatsd_origin_detection: false
 
+## @param dogstatsd_origin_detection_type - string - optional - default: ""
+## @env DD_DOGSTATSD_ORIGIN_DETECTION_TYPE - string - optional - default: ""
+## A DogStatsD client can send its container ID.
+## Set this parameter to "client" to instruct the Agent to use the container ID to enrich the metrics with container tags.
```
Comment (hush-hush) [inline pkg/config/config_template.yaml:1532]: Are we planning other value than `client` ? Could this be a boolean instead ?

**Review by hush-hush — CHANGES_REQUESTED** [review]
Looking good overall. We would need to update our official documentation [here](https://docs.datadoghq.com/developers/dogstatsd/datagram_shell?tab=metrics#dogstatsd-protocol-v11) to add a new section for version `1.2`.

File: pkg/dogstatsd/enrich.go [hunk 8]
```diff
@@ -39,25 +40,31 @@ func extractTagsMetadata(tags []string, defaultHostname string, originTags strin
...
 		// the tagger for entity that doesn't exist.
 
 		// currently only supported for pods
-		k8sOrigin = kubelet.KubePodTaggerEntityPrefix + entityIDValue
+		originFromClient = kubelet.KubePodTaggerEntityPrefix + originFromTag
+	} else if originFromTag == "" && len(originFromMsg) > 0 {
+		// originFromMsg is the container id sent by the newer clients.
+		// Opt-in is handled when parsing.
+		originFromClient = containers.BuildTaggerEntityName(string(originFromMsg))
 	}
```
Comment (clamoriniere) [inline pkg/dogstatsd/enrich.go:66]: global comment if you can put a matrix in the comment that explain who win if sever origin is provided it can help futur developer to understand 

```golang
| originFromTag | originFromUDS | originFromMsg | Result        |
|---------------|---------------|---------------|---------------|
| 0             | 0             | 0             | 0             |
| 0             | 0             | 1             |               |
| 0             | 1             | 0             | originFromUDS |
....


```

File: pkg/config/config_template.yaml (hunk 7 above)
Comment (ahmed-mez) [reply pkg/config/config_template.yaml:1607]: @clamoriniere any thoughts on this?

Comment (hush-hush) [reply pkg/config/config_template.yaml]: My main issue is that `dogstatsd_origin_detection_type: client` might imply that we will no longer support UDS origin detection while we would support both at the same time with the client side having priority.

Naming is always hard...
What about `dogstatsd_client_side_origin_detection: <bool>` ?

Or we keep the naming but make it a list of and deprecated `dogstatsd_origin_detection` ?
```
dogstatsd_origin_detection_type:
  - uds
  - client
  - ...
```
This will mean more code change in this PR and I'm not sure it's worth it.

WDYT ?

File: pkg/dogstatsd/enrich.go [hunk 9]
```diff
@@ -20,16 +21,41 @@ var (
 	CardinalityTagPrefix = "dd.internal.card:"
 )
 
-func extractTagsMetadata(tags []string, defaultHostname string, originTags string, entityIDPrecedenceEnabled bool) ([]string, string, string, string, string) {
+// extractTagsMetadata returns tags (client tags + host tag) and information needed to query tagger (origins, cardinality).
+//
+// The following tables explain how the origins are chosen.
+// originFromUDS is the origin discovered via UDS origin detection (container ID).
+// originFromTag is the origin sent by the client via the dd.internal.entity_id tag (pod uid).
```
Comment (clamoriniere) [inline pkg/dogstatsd/enrich.go:28]:
```suggestion
// originFromTag is the origin sent by the client via the dd.internal.entity_id tag (only support pod UID without prefix).
```

File: pkg/config/config_template.yaml (hunk 7 above)
Comment (ahmed-mez) [reply pkg/config/config_template.yaml]: Booleans are usually less error prone, so I'm leaning towards keeping it a bool - I renamed it `dogstatsd_origin_detection_client` in this [commit](https://github.com/DataDog/datadog-agent/pull/10659/commits/6148df977dd6c00573701c9d7c409c2774b6d7c1) . I can make it `dogstatsd_client_side_origin_detection` but it's a bit long

**Review by clamoriniere — APPROVED** [review]

File: releasenotes/notes/dsd-container-b2039d8e0e510fbf.yaml [hunk 10]
```diff
@@ -0,0 +1,5 @@
+---
+enhancements:
+  - |
+    The DogstatsD protocol now supports a new field to container the client's container ID.
+    This allows enriching DogstatsD metrics with container tags.
```
Comment (hush-hush) [inline releasenotes/notes/dsd-container-b2039d8e0e510fbf.yaml:5]:
```suggestion
    The DogstatsD protocol now supports a new field that contains the client's container ID.
    This allows enriching DogstatsD metrics with container tags.
```

**Review by hush-hush — APPROVED** [review]
Added a nit-pick on the release note. Looks good otherwise

File: pkg/dogstatsd/enrich.go [hunk 11]
```diff
@@ -39,25 +65,31 @@ func extractTagsMetadata(tags []string, defaultHostname string, originTags strin
 	}
 	tags = tags[:n]
 
-	origin := ""
+	udsOrigin := ""
 	// We use the UDS socket origin if no origin ID was specify in the tags
 	// or 'dogstatsd_entity_id_precedence' is set to False (default false).
-	if entityIDValue == "" || !entityIDPrecedenceEnabled {
+	if originFromTag == "" || !entityIDPrecedenceEnabled {
```
Comment (vickenty) [inline pkg/dogstatsd/enrich.go:71]: Should we skip `originFromUDS` if we are going to use `originFromMsg`? It would be nice to avoid tagging metrics with two container tags if a proxy is using UDS to send metrics on behalf of another container.

File: pkg/config/config_template.yaml [hunk 12]
```diff
@@ -1601,6 +1601,14 @@ api_key:
 #
 # dogstatsd_origin_detection: false
 
+## @param dogstatsd_origin_detection_client - boolean - optional - default: false
+## @env DD_DOGSTATSD_ORIGIN_DETECTION_CLIENT - boolean - optional - default: false
+## A DogStatsD client can send its container ID.
+## Enable this parameter to instruct the Agent to use the container ID to enrich the metrics with container tags.
```
Comment (vickenty) [inline pkg/config/config_template.yaml:1607]: Would this make sense?
```suggestion
## Whether the Agent should use a client-provided container ID to enrich the metrics, events and service checks with container tags.
```

Would it make sense to mention that this affects dogstatsd_tag_cardinality above `low` here?

**Review by vickenty — COMMENTED** [review]
Would it make sense to c

File: pkg/dogstatsd/enrich.go (hunk 11 above)
Comment (ahmed-mez) [reply pkg/dogstatsd/enrich.go:71]: UDS origin detection is opt-in and not enabled by default, I'd rather let the user have control.

File: pkg/config/config_template.yaml (hunk 12 above)
Comment (ahmed-mez) [reply pkg/config/config_template.yaml:1607]: > Would it make sense to mention that this affects dogstatsd_tag_cardinality above low here?

I'm not sure I understand, mind clarifying ?

File: pkg/dogstatsd/enrich.go (hunk 11 above)
Comment (vickenty) [reply pkg/dogstatsd/enrich.go:71]: This feature would probably interest users that use origin detection today, so they may be inclined to have both options enabled: use uds to tag traffic on older clients, while using benefits of client-side origin detection with newer clients. Unless double-tagging is a use case we explicitly want to support, I think it would make more sense to assign only one origin per metric.

File: pkg/config/config_template.yaml (hunk 12 above)
Comment (vickenty) [reply pkg/config/config_template.yaml]: Nevermind, I got confused and thought `low` cardinality tags are somehow not derived from the origin.

File: pkg/dogstatsd/enrich.go (hunk 11 above)
Comment (ahmed-mez) [reply pkg/dogstatsd/enrich.go:71]: I'd rather keep it simple as is currently and not make any assumptions.
We agreed that extending the protocol is always an option in the future, so if we want to turn on and off UDS origin detection on-demand it would be best to let the client do it via a new field.
